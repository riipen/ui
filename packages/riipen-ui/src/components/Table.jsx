import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";
import withForwardedRef from "../utils/withForwardedRef";

class Table extends React.Component {
  static propTypes = {
    /**
     * What background to render for the table.
     * Defaults to grey100.
     */
    backgroundColor: PropTypes.oneOf([
      "transparent",
      "grey50",
      "grey100",
      "grey200",
      "grey300",
      "grey400"
    ]),

    /**
     * Children to render in the table element.
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element.
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * A reference to the table DOM element.
     */
    forwardedRef: PropTypes.element,

    /**
     * Whether to give all columns in the table equal width.
     */
    layout: PropTypes.oneOf(["auto", "fixed"])
  };

  static defaultProps = {
    backgroundColor: "grey100",
    layout: "auto"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      layout,
      backgroundColor,
      forwardedRef
    } = this.props;

    const theme = this.context;

    const tableBackground = {
      transparent: "transparent",
      grey50: theme.palette.grey[50],
      grey100: theme.palette.grey[100],
      grey200: theme.palette.grey[200],
      grey300: theme.palette.grey[300],
      grey400: theme.palette.grey[400]
    };

    return (
      <React.Fragment>
        <table ref={forwardedRef} className={clsx(classes)} border="0">
          {children}
        </table>
        <style jsx>{`
          table {
            background-color: ${tableBackground[backgroundColor]};
            border-radius: 4px;
            border-collapse: collapse;
            table-layout: ${layout};
            width: 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withForwardedRef(withClasses(Table));
