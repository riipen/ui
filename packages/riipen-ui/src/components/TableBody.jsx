import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

class TableBody extends React.Component {
  static propTypes = {
    /**
     * Children to render in the tbody element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string)
  };

  render() {
    const { children, classes, ...other } = this.props;
    return (
      <tbody classes={clsx(classes)} {...other}>
        {children}
      </tbody>
    );
  }
}

export default withClasses(TableBody);
