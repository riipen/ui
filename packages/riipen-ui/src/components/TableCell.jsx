import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class TableCell extends React.Component {
  static propTypes = {
    /**
     * Alignment property to apply to the td element
     */
    align: PropTypes.oneOf(["center", "left", "right", "justify", "char"]),

    /**
     * Children to render in the td element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the td element
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * Number of columns the td element should span
     */
    colSpan: PropTypes.number,

    /**
     * The amount of padding the cell should have.
     */
    padding: PropTypes.number,

    /**
     * Number of rows the td element should span
     */
    rowSpan: PropTypes.number
  };

  static defaultProps = {
    align: "left",
    colSpan: 1,
    padding: 3,
    rowSpan: 1
  };

  static contextType = ThemeContext;

  render() {
    const {
      align,
      children,
      classes,
      colSpan,
      padding,
      rowSpan,
      ...other
    } = this.props;

    const theme = this.context;

    return (
      <React.Fragment>
        <td
          className={clsx(classes)}
          align={align}
          colSpan={colSpan}
          rowSpan={rowSpan}
          {...other}
        >
          {children}
        </td>
        <style jsx>{`
          td {
            font-family: ${theme.typography.body2.fontFamily};
            font-size: ${theme.typography.body2.fontSize};
            font-weight: ${theme.typography.body2.fontWeight};
            letter-spacing: ${theme.typography.body2.letterSpacing};
            line-height: ${theme.typography.body2.lineHeight};
            padding: ${theme.spacing(padding)}px;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses(TableCell);
