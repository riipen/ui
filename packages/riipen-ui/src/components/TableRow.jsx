import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class TableRow extends React.Component {
  static propTypes = {
    /**
     * Children to render in the tr element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * Whether to highlight the row on hover
     */
    hover: PropTypes.bool,

    /**
     * Whether to render the border for the row
     */
    border: PropTypes.bool
  };

  static defaultProps = {
    border: true
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, hover, border } = this.props;

    const theme = this.context;

    return (
      <React.Fragment>
        <tr className={clsx(classes)}>{children}</tr>
        <style jsx>{`
          tr {
            border-bottom: 1px solid
              ${border &&
                (theme.palette.divider
                  ? theme.palette.divider
                  : `transparent`)};
            border-left: 3px solid transparent;
          }

          tr:hover {
            ${hover ? `background-color: ${theme.palette.common.white};` : ""}
            ${hover
              ? `border-left: 3px solid ${theme.palette.secondary.main};`
              : ""}
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(TableRow);
