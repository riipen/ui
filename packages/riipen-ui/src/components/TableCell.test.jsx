import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import defaultTheme from "../themes/default";

import TableCell from "./TableCell";
import Table from "./Table";
import TableBody from "./TableBody";
import TableRow from "./TableRow";

describe("<TableCell>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell />
            </TableRow>
          </TableBody>
        </Table>
      );
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(
      <Table>
        <TableBody>
          <TableRow>
            <TableCell />
          </TableRow>
        </TableBody>
      </Table>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new TableCell().type.defaultProps;

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(wrapper.find("TableCell").props().colSpan).toEqual(
        defaultProps.colSpan
      );
      expect(wrapper.find("TableCell").props().padding).toEqual(
        defaultProps.padding
      );
      expect(wrapper.find("TableCell").props().rowSpan).toEqual(
        defaultProps.rowSpan
      );
      expect(wrapper.find("TableCell").props().textAlign).toEqual(
        defaultProps.textAlign
      );
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>child</div>;

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell>{child}</TableCell>
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell classes={classes} />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(
        wrapper
          .find("TableCell")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
      const classes = ["riipen", "riipen-tablecell"];

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(
        wrapper
          .find("TableCell")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("colSpan prop", () => {
    it("sets custom colSpan", () => {
      const colSpan = 2;

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell colSpan={colSpan} />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(
        wrapper
          .find("TableCell")
          .find("td")
          .props().colSpan
      ).toEqual(colSpan);
    });
  });

  describe("padding prop", () => {
    it("sets custom padding", () => {
      const padding = 2;

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell padding={padding} />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(
        wrapper
          .find("TableCell")
          .childAt(0)
          .props().style.padding
      ).toEqual(`${defaultTheme.spacing(padding)}px`);
    });
  });

  describe("rowSpan prop", () => {
    it("sets custom rowSpan", () => {
      const rowSpan = 2;

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell rowSpan={rowSpan} />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(
        wrapper
          .find("TableCell")
          .find("td")
          .props().rowSpan
      ).toEqual(rowSpan);
    });
  });

  describe("textAlign prop", () => {
    it("sets custom valid textAlign", () => {
      const textAlign = "center";

      const wrapper = mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell textAlign={textAlign} />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(
        wrapper
          .find("TableCell")
          .childAt(0)
          .props().style.textAlign
      ).toEqual(textAlign);
    });

    it("gives an error when given an invalid textAlign", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const textAlign = "centered";

      mount(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell textAlign={textAlign} />
            </TableRow>
          </TableBody>
        </Table>
      );

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
