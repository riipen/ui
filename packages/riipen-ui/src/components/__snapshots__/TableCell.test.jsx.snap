// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableCell> renders correct snapshot 1`] = `
<ForwardRef(withForwardedRef(withClasses(Table)))>
  <withClasses(Table)
    forwardedRef={null}
  >
    <Table
      backgroundColor="grey50"
      classes={
        Array [
          "riipen",
          "riipen-table",
        ]
      }
      forwardedRef={null}
      layout="auto"
    >
      <table
        border="0"
        className="jsx-1059775881 riipen riipen-table"
      >
        <withClasses(TableBody)>
          <TableBody
            classes={
              Array [
                "riipen",
                "riipen-tablebody",
              ]
            }
          >
            <tbody
              classes="riipen riipen-tablebody"
            >
              <withClasses(TableRow)>
                <TableRow
                  border={true}
                  classes={
                    Array [
                      "riipen",
                      "riipen-tablerow",
                    ]
                  }
                  forceHover={false}
                >
                  <tr
                    className="jsx-4111142428 riipen riipen-tablerow"
                  >
                    <withClasses(TableCell)>
                      <TableCell
                        classes={
                          Array [
                            "riipen",
                            "riipen-tablecell",
                          ]
                        }
                        colSpan={1}
                        padding={3}
                        rowSpan={1}
                        textAlign="left"
                      >
                        <td
                          className="jsx-3771758953 riipen riipen-tablecell"
                          colSpan={1}
                          rowSpan={1}
                        />
                        <JSXStyle
                          dynamic={
                            Array [
                              "Roboto, Helvetica, Arial, sans-serif",
                              "14px",
                              400,
                              "0.01071em",
                              "20px",
                              12,
                              "left",
                            ]
                          }
                          id="498164393"
                        />
                      </TableCell>
                    </withClasses(TableCell)>
                  </tr>
                  <JSXStyle
                    dynamic={
                      Array [
                        "#ddd",
                        "#fff",
                        "#f58632",
                        "",
                        "",
                      ]
                    }
                    id="2808869914"
                  />
                </TableRow>
              </withClasses(TableRow)>
            </tbody>
          </TableBody>
        </withClasses(TableBody)>
      </table>
      <JSXStyle
        dynamic={
          Array [
            "#fafafa",
            "auto",
          ]
        }
        id="4123435624"
      />
    </Table>
  </withClasses(Table)>
</ForwardRef(withForwardedRef(withClasses(Table)))>
`;
